File: app/src/main/java/ru/e6atb/chat/SessionStore.java
package ru.e6atb.chat;

import android.content.Context;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.security.MessageDigest;
import java.util.Properties;

import rs.ove.crypt.proto.NativeE2E;

final class SessionStore {
	private static final String FILE_NAME = "e6atb.session.properties";
	private static final String SERVER = "server";
	private static final String TOKEN = "token";
	private static final String USER_ID = "user_id";
	private static final String USERNAME = "username";
	private static final String LEGACY_LOGIN = "login";
	private static final String LAST_UPDATE = "last_update";
	private static final String BACKGROUND_LAST_UPDATE = "background_last_update";
	private static final String NOTIFICATION_BOOTSTRAP_COMPLETE = "notification_bootstrap_complete";
	private static final String LAST_GITHUB_UPDATE_CHECK_AT = "last_github_update_check_at";
	private static final String SHOW_STATUS = "show_status";
	private static final String USE_INSETS = "use_insets";
	private static final String LANGUAGE = "language";
	private static final String TRANSPORT_PROTOCOL = "transport_protocol";
	static final String TRANSPORT_AUTO = "auto";
	static final String TRANSPORT_MST5 = "mst5";
	static final String TRANSPORT_M5OH = "m5oh";
	private static final String E2E_PEER_PREFIX = "e2e.peer.";

	private SessionStore() {
	}

	static void save(Context context, String server, String token, String username) {
		save(context, server, token, userId(context), username);
	}

	static void save(Context context, String server, String token, String userId, String username) {
		Properties p = load(context);
		p.setProperty(SERVER, normalizeServer(server));
		p.setProperty(TOKEN, safe(token));
		p.setProperty(USER_ID, safe(userId));
		p.setProperty(USERNAME, safe(username));
		p.remove(LEGACY_LOGIN);
		store(context, p);
	}

	static void saveServer(Context context, String server) {
		Properties p = load(context);
		p.setProperty(SERVER, normalizeServer(server));
		store(context, p);
	}

	static void clear(Context context) {
		Properties p = load(context);
		p.remove(TOKEN);
		p.remove(USER_ID);
		p.remove(USERNAME);
		p.remove(LEGACY_LOGIN);
		p.remove(LAST_UPDATE);
		p.remove(BACKGROUND_LAST_UPDATE);
		p.remove(NOTIFICATION_BOOTSTRAP_COMPLETE);
		store(context, p);
	}

	static boolean hasSession(Context context) {
		return token(context).length() > 0;
	}

	static String server(Context context, String fallback) {
		Properties properties = load(context);
		// The production endpoint is fixed; this also migrates old arbitrary and
		// legacy server choices without affecting the user's authentication token.
		String fixed = normalizeServer(fallback);
		if (!fixed.equals(normalizeServer(properties.getProperty(SERVER, "")))) {
			properties.setProperty(SERVER, fixed);
			store(context, properties);
		}
		return fixed;
	}

	static String transportProtocol(Context context) {
		String protocol = get(context, TRANSPORT_PROTOCOL, TRANSPORT_AUTO);
		if (TRANSPORT_MST5.equals(protocol) || TRANSPORT_M5OH.equals(protocol)) return protocol;
		return TRANSPORT_AUTO;
	}

	static void setTransportProtocol(Context context, String protocol) {
		Properties p = load(context);
		if (!TRANSPORT_MST5.equals(protocol) && !TRANSPORT_M5OH.equals(protocol)) {
			protocol = TRANSPORT_AUTO;
		}
		p.setProperty(TRANSPORT_PROTOCOL, protocol);
		store(context, p);
	}

	/**
	 * Resolves the endpoint used for a connection without changing the stable
	 * server key that scopes cached chats and queued messages.
	 */
	static String transportEndpoint(Context context, String fallback) {
		String endpoint = server(context, fallback);
		String protocol = transportProtocol(context);
		if (TRANSPORT_AUTO.equals(protocol)) return endpoint;
		String prefix = TRANSPORT_MST5.equals(protocol) ? "mst5://" : "m5oh://";
		for (String candidate : endpoint.split(java.util.regex.Pattern.quote("|"))) {
			if (candidate.trim().toLowerCase(java.util.Locale.US).startsWith(prefix)) {
				return candidate.trim();
			}
		}
		// A fallback without the requested transport is still usable.
		return endpoint;
	}

	static String normalizeServer(String server) {
		String s = safe(server).trim();
		while (s.endsWith("/")) {
			s = s.substring(0, s.length() - 1);
		}
		String lower = s.toLowerCase(java.util.Locale.US);
		String legacyPrefix = "tcp" + "://";
		if (lower.startsWith(legacyPrefix)) {
			return s.substring(legacyPrefix.length());
		}
		return s;
	}

	static String token(Context context) {
		return get(context, TOKEN, "");
	}

	static String login(Context context) {
		String username = get(context, USERNAME, "");
		return username.length() > 0 ? username : get(context, LEGACY_LOGIN, "");
	}

	static String userId(Context context) {
		return get(context, USER_ID, "");
	}

	static long lastUpdate(Context context) {
		return longValue(context, LAST_UPDATE);
	}

	static void lastUpdate(Context context, long id) {
		setLong(context, LAST_UPDATE, id);
	}

	static long backgroundLastUpdate(Context context) {
		return longValue(context, BACKGROUND_LAST_UPDATE);
	}

	static void backgroundLastUpdate(Context context, long id) {
		setLong(context, BACKGROUND_LAST_UPDATE, id);
	}

	static boolean notificationBootstrapComplete(Context context) {
		// Existing installations predate this flag and already have a valid
		// cursor, so default them to ready. A fresh login explicitly stores
		// false before either polling loop is started.
		return getBoolean(context, NOTIFICATION_BOOTSTRAP_COMPLETE, true);
	}

	static void notificationBootstrapComplete(Context context, boolean complete) {
		setBoolean(context, NOTIFICATION_BOOTSTRAP_COMPLETE, complete);
	}

	static long lastGithubUpdateCheckAt(Context context) {
		return longValue(context, LAST_GITHUB_UPDATE_CHECK_AT);
	}

	static void lastGithubUpdateCheckAt(Context context, long timestamp) {
		setLong(context, LAST_GITHUB_UPDATE_CHECK_AT, timestamp);
	}

	private static long longValue(Context context, String key) {
		String s = get(context, key, "0");
		try {
			return Long.parseLong(s);
		} catch (Exception e) {
			return 0;
		}
	}

	private static void setLong(Context context, String key, long id) {
		Properties p = load(context);
		p.setProperty(key, String.valueOf(id));
		store(context, p);
	}

	static boolean showStatus(Context context) {
		return getBoolean(context, SHOW_STATUS, true);
	}

	static void showStatus(Context context, boolean enabled) {
		setBoolean(context, SHOW_STATUS, enabled);
	}

	static boolean useInsets(Context context) {
		return getBoolean(context, USE_INSETS, android.os.Build.VERSION.SDK_INT >= 20);
	}

	static void useInsets(Context context, boolean enabled) {
		setBoolean(context, USE_INSETS, enabled);
	}

	static String language(Context context) {
		return get(context, LANGUAGE, AppLocale.SYSTEM);
	}

	static void language(Context context, String language) {
		Properties p = load(context);
		p.setProperty(LANGUAGE, safe(language));
		store(context, p);
	}

	static NativeE2E.Identity e2eIdentity(Context context, String login) {
		try {
			return NativeE2E.open(context, keyID(login), false);
		} catch (Exception ignored) {
			return null;
		}
	}

	static void clearLegacyE2EIdentity(Context context, String login) {
		Properties p = load(context);
		p.remove("e2e.private." + keyID(login));
		store(context, p);
	}

	static NativeE2E.Identity createE2EIdentity(Context context, String login) {
		try {
			return NativeE2E.open(context, keyID(login), true);
		} catch (Exception error) {
			throw new IllegalStateException("cannot create native E2E identity", error);
		}
	}

	static void saveE2EIdentity(Context context, String login, NativeE2E.Identity identity) {
		// NativeE2E restore/save is atomic; managed code never receives private key bytes.
	}

	static void clearE2EIdentity(Context context, String login) {
		try {
			NativeE2E.Identity identity = e2eIdentity(context, login);
			if (identity != null) identity.remove();
		} catch (Exception ignored) {
		}
		Properties p = load(context);
		// Remove legacy Java private keys after the incompatible v3 migration.
		p.remove("e2e.private." + keyID(login));
		store(context, p);
	}

	static boolean pinPeerE2EKey(Context context, String server, String ownLogin, String peer, String publicKey) {
		String key = E2E_PEER_PREFIX + keyID(server + "\n" + ownLogin + "\n" + peer);
		Properties p = load(context);
		String current = p.getProperty(key);
		if (current == null || !current.equals(publicKey)) {
			p.setProperty(key, publicKey);
			store(context, p);
			return current != null;
		}
		return false;
	}

	static String peerE2EFingerprint(Context context, String server, String ownLogin, String peer) {
		String key = E2E_PEER_PREFIX + keyID(server + "\n" + ownLogin + "\n" + peer);
		String publicKey = get(context, key, "");
		if (publicKey.length() == 0) return "";
		try {
			return NativeE2E.fingerprint(publicKey);
		} catch (Exception ignored) {
			return "";
		}
	}

	private static boolean getBoolean(Context context, String key, boolean fallback) {
		String value = get(context, key, fallback ? "true" : "false");
		return "true".equals(value);
	}

	private static void setBoolean(Context context, String key, boolean value) {
		Properties p = load(context);
		p.setProperty(key, value ? "true" : "false");
		store(context, p);
	}

	private static String get(Context context, String key, String fallback) {
		Properties p = load(context);
		String value = p.getProperty(key);
		return value == null ? fallback : value;
	}

	private static synchronized Properties load(Context context) {
		Properties p = new Properties();
		File f = file(context);
		if (f == null || !f.exists()) {
			return p;
		}

		FileInputStream in = null;
		try {
			in = new FileInputStream(f);
			p.load(in);
		} catch (Exception ignored) {
		} finally {
			closeQuietly(in);
		}

		return p;
	}

	private static synchronized void store(Context context, Properties p) {
		File f = file(context);
		if (f == null) {
			return;
		}

		File parent = f.getParentFile();
		if (parent != null && !parent.exists()) {
			parent.mkdirs();
		}

		FileOutputStream out = null;
		try {
			out = new FileOutputStream(f);
			p.store(out, "");
			out.flush();
		} catch (Exception ignored) {
		} finally {
			closeQuietly(out);
		}
	}

	private static File file(Context context) {
		if (context == null) {
			return null;
		}

		Context app = context.getApplicationContext();
		if (app == null) {
			app = context;
		}

		File dir = app.getFilesDir();
		if (dir == null) {
			return null;
		}

		if (!dir.exists()) {
			dir.mkdirs();
		}

		return new File(dir, FILE_NAME);
	}

	private static void closeQuietly(java.io.Closeable c) {
		if (c == null) {
			return;
		}

		try {
			c.close();
		} catch (Exception ignored) {
		}
	}

	private static String safe(String s) {
		return s == null ? "" : s;
	}

	private static String keyID(String value) {
		try {
			byte[] hash = MessageDigest.getInstance("SHA-256").digest(safe(value).getBytes("UTF-8"));
			StringBuilder out = new StringBuilder(hash.length * 2);
			for (int i = 0; i < hash.length; i++) {
				int b = hash[i] & 0xff;
				if (b < 16) out.append('0');
				out.append(Integer.toHexString(b));
			}
			return out.toString();
		} catch (Exception ex) {
			throw new IllegalStateException(ex);
		}
	}
}

File: app/src/main/kotlin/ru/e6atb/chat/SessionStore.kt
package ru.e6atb.chat

import android.content.Context
import android.os.Build
import org.json.JSONObject
import rs.ove.crypt.proto.NativeE2E
import rs.ove.crypt.proto.NativeMst5
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.security.MessageDigest
import java.util.Locale
import java.util.Properties

internal object SessionStore {
    private const val FILE_NAME = "e6atb.session.properties"
    private const val SERVER = "server"
    private const val TOKEN = "token"
    private const val USER_ID = "user_id"
    private const val USERNAME = "username"
    private const val LEGACY_LOGIN = "login"
    private const val LAST_UPDATE = "last_update"
    private const val BACKGROUND_LAST_UPDATE = "background_last_update"
    private const val NOTIFICATION_BOOTSTRAP_COMPLETE = "notification_bootstrap_complete"
    private const val LAST_GITHUB_UPDATE_CHECK_AT = "last_github_update_check_at"
    private const val SHOW_STATUS = "show_status"
    private const val USE_INSETS = "use_insets"
    private const val LANGUAGE = "language"
    private const val TRANSPORT_PROTOCOL = "transport_protocol"
    @JvmField val TRANSPORT_AUTO = "auto"
    @JvmField val TRANSPORT_MST5 = "mst5"
    @JvmField val TRANSPORT_M5OH = "m5oh"
    private const val E2E_PEER_PREFIX = "e2e.peer."

    @JvmStatic fun save(context: Context?, server: String?, token: String?, username: String?) = save(context, server, token, userId(context), username)
    @JvmStatic fun save(context: Context?, server: String?, token: String?, userId: String?, username: String?) {
        val p = load(context)
        p.setProperty(SERVER, normalizeServer(server)); p.setProperty(TOKEN, safe(token)); p.setProperty(USER_ID, safe(userId)); p.setProperty(USERNAME, safe(username)); p.remove(LEGACY_LOGIN)
        store(context, p)
    }
    @JvmStatic fun saveServer(context: Context?, server: String?) { load(context).also { it.setProperty(SERVER, normalizeServer(server)); store(context, it) } }
    @JvmStatic fun clear(context: Context?) { load(context).also { p -> listOf(TOKEN, USER_ID, USERNAME, LEGACY_LOGIN, LAST_UPDATE, BACKGROUND_LAST_UPDATE, NOTIFICATION_BOOTSTRAP_COMPLETE).forEach(p::remove); store(context, p) } }
    @JvmStatic fun hasSession(context: Context?): Boolean = token(context).isNotEmpty()
    @JvmStatic fun server(context: Context?, fallback: String?): String {
        val properties = load(context); val fixed = normalizeServer(fallback)
        if (fixed != normalizeServer(properties.getProperty(SERVER, ""))) { properties.setProperty(SERVER, fixed); store(context, properties) }
        return fixed
    }
    @JvmStatic fun transportProtocol(context: Context?): String = get(context, TRANSPORT_PROTOCOL, TRANSPORT_AUTO).let { if (it == TRANSPORT_MST5 || it == TRANSPORT_M5OH) it else TRANSPORT_AUTO }
    @JvmStatic fun setTransportProtocol(context: Context?, protocol: String?) { load(context).also { p -> p.setProperty(TRANSPORT_PROTOCOL, if (protocol == TRANSPORT_MST5 || protocol == TRANSPORT_M5OH) protocol else TRANSPORT_AUTO); store(context, p) } }
    @JvmStatic fun transportEndpoint(context: Context?, fallback: String?): String = server(context, fallback)
    @JvmStatic fun normalizeServer(server: String?): String {
        var value = safe(server).trim(); while (value.endsWith('/')) value = value.dropLast(1)
        return if (value.lowercase(Locale.US).startsWith("tcp://")) value.substring(6) else value
    }
    @JvmStatic fun token(context: Context?): String = get(context, TOKEN, "")
    @JvmStatic fun login(context: Context?): String = get(context, USERNAME, "").let { if (it.isNotEmpty()) it else get(context, LEGACY_LOGIN, "") }
    @JvmStatic fun userId(context: Context?): String = get(context, USER_ID, "")
    @JvmStatic fun lastUpdate(context: Context?): Long = longValue(context, LAST_UPDATE)
    @JvmStatic fun lastUpdate(context: Context?, id: Long) = setLong(context, LAST_UPDATE, id)
    @JvmStatic fun backgroundLastUpdate(context: Context?): Long = longValue(context, BACKGROUND_LAST_UPDATE)
    @JvmStatic fun backgroundLastUpdate(context: Context?, id: Long) = setLong(context, BACKGROUND_LAST_UPDATE, id)
    @JvmStatic fun notificationBootstrapComplete(context: Context?): Boolean = getBoolean(context, NOTIFICATION_BOOTSTRAP_COMPLETE, true)
    @JvmStatic fun notificationBootstrapComplete(context: Context?, complete: Boolean) = setBoolean(context, NOTIFICATION_BOOTSTRAP_COMPLETE, complete)
    @JvmStatic fun lastGithubUpdateCheckAt(context: Context?): Long = longValue(context, LAST_GITHUB_UPDATE_CHECK_AT)
    @JvmStatic fun lastGithubUpdateCheckAt(context: Context?, timestamp: Long) = setLong(context, LAST_GITHUB_UPDATE_CHECK_AT, timestamp)
    @JvmStatic fun showStatus(context: Context?): Boolean = getBoolean(context, SHOW_STATUS, true)
    @JvmStatic fun showStatus(context: Context?, enabled: Boolean) = setBoolean(context, SHOW_STATUS, enabled)
    @JvmStatic fun useInsets(context: Context?): Boolean = getBoolean(context, USE_INSETS, Build.VERSION.SDK_INT >= 20)
    @JvmStatic fun useInsets(context: Context?, enabled: Boolean) = setBoolean(context, USE_INSETS, enabled)
    @JvmStatic fun language(context: Context?): String = get(context, LANGUAGE, AppLocale.SYSTEM)
    @JvmStatic fun language(context: Context?, language: String?) { load(context).also { p -> p.setProperty(LANGUAGE, safe(language)); store(context, p) } }

    @JvmStatic fun e2eIdentity(context: Context?, login: String?): NativeE2E.Identity? = try { NativeE2E.open(context, keyID(login), false) } catch (_: Exception) { null }
    @JvmStatic fun clearLegacyE2EIdentity(context: Context?, login: String?) { load(context).also { p -> p.remove("e2e.private.${keyID(login)}"); store(context, p) } }
    @JvmStatic fun createE2EIdentity(context: Context?, login: String?): NativeE2E.Identity = try { NativeE2E.open(context, keyID(login), true) } catch (error: Exception) { throw IllegalStateException("cannot create native E2E identity", error) }
    @JvmStatic fun clearE2EIdentity(context: Context?, login: String?) {
        try { e2eIdentity(context, login)?.remove() } catch (_: Exception) { }
        clearLegacyE2EIdentity(context, login)
    }
    @JvmStatic fun pinPeerE2EKey(context: Context?, server: String?, ownLogin: String?, peer: String?, publicKey: String?): Boolean {
        val key = E2E_PEER_PREFIX + keyID("${safe(server)}\n${safe(ownLogin)}\n${safe(peer)}")
        val p = load(context); val current = p.getProperty(key)
        if (current == null || current != publicKey) { p.setProperty(key, safe(publicKey)); store(context, p); return current != null }; return false
    }
    @JvmStatic fun peerE2EFingerprint(context: Context?, server: String?, ownLogin: String?, peer: String?): String {
        val key = E2E_PEER_PREFIX + keyID("${safe(server)}\n${safe(ownLogin)}\n${safe(peer)}"); val publicKey = get(context, key, "")
        return if (publicKey.isEmpty()) "" else try { NativeE2E.fingerprint(publicKey) } catch (_: Exception) { "" }
    }

    private fun longValue(context: Context?, key: String): Long = get(context, key, "0").toLongOrNull() ?: 0
    private fun setLong(context: Context?, key: String, value: Long) { load(context).also { p -> p.setProperty(key, value.toString()); store(context, p) } }
    private fun getBoolean(context: Context?, key: String, fallback: Boolean): Boolean = get(context, key, fallback.toString()) == "true"
    private fun setBoolean(context: Context?, key: String, value: Boolean) { load(context).also { p -> p.setProperty(key, value.toString()); store(context, p) } }
    private fun get(context: Context?, key: String, fallback: String): String = load(context).getProperty(key) ?: fallback
    @Synchronized private fun load(context: Context?): Properties = try {
        val snapshot = JSONObject(NativeMst5.sessionSnapshot())
        if (snapshot.optBoolean("exists", false)) properties(snapshot.optJSONObject("values")) else loadLegacy(context).also { if (it.isNotEmpty()) store(context, it) }
    } catch (_: Exception) { loadLegacy(context) }
    @Synchronized private fun loadLegacy(context: Context?): Properties {
        val properties = Properties(); val file = file(context) ?: return properties; if (!file.exists()) return properties
        try { FileInputStream(file).use(properties::load) } catch (_: Exception) { }; return properties
    }
    @Synchronized private fun store(context: Context?, properties: Properties) { try { NativeMst5.replaceSession(propertiesJson(properties).toString()); return } catch (_: Exception) { }; storeLegacy(context, properties) }
    @Synchronized private fun storeLegacy(context: Context?, properties: Properties) {
        val file = file(context) ?: return; file.parentFile?.let { if (!it.exists()) it.mkdirs() }
        try { FileOutputStream(file).use { properties.store(it, ""); it.flush() } } catch (_: Exception) { }
    }
    private fun properties(values: JSONObject?): Properties = Properties().also { out -> values?.keys()?.forEach { key -> values.opt(key).takeUnless { it == null || it == JSONObject.NULL }?.let { out.setProperty(key, it.toString()) } } }
    private fun propertiesJson(values: Properties): JSONObject = JSONObject().also { out -> values.stringPropertyNames().forEach { out.put(it, values.getProperty(it, "")) } }
    private fun file(context: Context?): File? { val app = context?.applicationContext ?: context ?: return null; val directory = app.filesDir ?: return null; if (!directory.exists()) directory.mkdirs(); return File(directory, FILE_NAME) }
    private fun safe(value: String?): String = value ?: ""
    private fun keyID(value: String?): String = try { MessageDigest.getInstance("SHA-256").digest(safe(value).toByteArray(Charsets.UTF_8)).joinToString("") { "%02x".format(it) } } catch (error: Exception) { throw IllegalStateException(error) }
}
